# Module 1A: Data Exploration & Preparation
## Housing Dataset

**Цели:**
- Зареждане и разглеждане на данни
- Exploratory Data Analysis (EDA)
- Обработка на категориални променливи
- Feature Scaling
- Train/Test Split

---

## 📚 REFERENCE SHEET

### Pandas Operations:
- `pd.read_csv(filename)` - зарежда CSV файл
- `df.head(n)` - показва първи n реда
- `df.tail(n)` - показва последни n реда
- `df.shape` - връща (rows, columns)
- `df.info()` - информация за колони и типове
- `df.dtypes` - типове данни на колоните
- `df.describe()` - статистика за числови колони
- `df.isnull()` - проверка за missing values
- `df.isnull().sum()` - брой missing values по колони
- `df.unique()` - уникални стойности
- `df.value_counts()` - брой на всяка уникална стойност
- `df.copy()` - прави копие на DataFrame
- `df.drop(columns, axis=1)` - премахва колони
- `df.map(dictionary)` - замества стойности по речник
- `pd.get_dummies(df, columns)` - One-Hot Encoding
- `df.columns` - имена на колоните
- `df.columns.tolist()` - имена като списък

### Matplotlib/Seaborn:
- `plt.figure(figsize=(w, h))` - създава нова фигура
- `plt.hist(data, bins)` - histogram
- `plt.scatter(x, y)` - scatter plot
- `plt.xlabel(label)` - етикет на X оста
- `plt.ylabel(label)` - етикет на Y оста
- `plt.title(title)` - заглавие
- `plt.show()` - показва графиката
- `df.boxplot(column, by)` - boxplot
- `df.corr()` - correlation matrix
- `sns.heatmap(data, annot=True)` - heatmap

### Sklearn:
- `StandardScaler()` - създава scaler обект
- `scaler.fit(X)` - изчислява mean и std
- `scaler.transform(X)` - прилага scaling
- `scaler.fit_transform(X)` - fit + transform заедно
- `train_test_split(X, y, test_size, random_state)` - разделя данни

### Numpy:
- `np.sqrt(x)` - квадратен корен
- `np.mean(x)` - средна стойност
- `np.std(x)` - стандартно отклонение

---

## Част 1: Зареждане на данните (5 минути)

In [5]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [6]:
# Задача 1.1: Импортирайте необходимите библиотеки
import pandas as pd       # за работа с таблици и данни
import numpy as np        # за числови операции и масиви
import matplotlib.pyplot as plt   # за графики
import seaborn as sns     # за по-красиви визуализации

**Защо използваме `as`:**
- Създаваме кратки псевдоними (aliases) за по-лесно писане
- `pd.read_csv()` е по-кратко от `pandas.read_csv()`
- Това е стандартна конвенция в data science

In [1]:
import os
path = "/content/drive/MyDrive/Machine Learning/Housing.csv"  # твоят път
print("Exists:", os.path.exists(path))

Exists: False


In [7]:
# Задача 1.2: Заредете 'Housing.csv' файла в променлива df; Пример - /content/drive/MyDrive/...../...../Housing.csv
df = pd.read_csv('/content/drive/MyDrive/Machine Learning/Housing.csv')

Какво прави кодът:

pd.read_csv() чете CSV файл и създава DataFrame
DataFrame е таблица с редове и колони (като Excel sheet)
Записваме данните в променлива df (стандартно име за DataFrame)
Защо използваме DataFrame:

Лесно манипулиране на данни
Вградени функции за анализ
Може да съдържа различни типове данни (числа, текст, дати)

In [8]:
# Задача 1.3: Покажете първите 5 реда
df.head()

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished


**Какво прави кодът:**
- Показва първите n реда (по подразбиране 5)
- Дава бърз преглед на структурата на данните

**Защо го правим:**
- Винаги първа стъпка след зареждане на данни
- Проверяваме дали данните са заредени правилно
- Виждаме имената на колоните и типовете стойности

**Очакван резултат:**
- Таблица с 5 реда и 13 колони
- Колони: price, area, bedrooms, bathrooms, stories, mainroad, guestroom, basement, hotwaterheating, airconditioning, parking, prefarea, furnishingstatus
- Някои колони с числа, други с 'yes'/'no', furnishingstatus с 'furnished'/'semi-furnished'/'unfurnished'

In [9]:
# Задача 1.4: Покажете последните 3 реда
df.tail(3)

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
542,1750000,3620,2,1,1,yes,no,no,no,no,0,no,unfurnished
543,1750000,2910,3,1,1,no,no,no,no,no,0,no,furnished
544,1750000,3850,3,1,2,yes,no,no,no,no,0,no,unfurnished


In [10]:
# Задача 1.5: Проверете размера на данните (брой редове и колони)
print(f"Размер: {df.shape}")

Размер: (545, 13)


**Какво прави кодът:**
- Връща tuple с (редове, колони)
- f-string форматира изхода за четимост

**Защо го правим:**
- Бърза проверка на размера на dataset-а
- Важно за разбиране на обема на данните
- Помага при debug (ако загубим редове/колони)

**Очакван резултат:**
- "Размер: (545, 13)"
- 545 къщи (observations)
- 13 характеристики (features)

In [11]:
# Задача 1.6: Покажете информация за колоните и типовете данни
_____

NameError: name '_____' is not defined

**Какво прави кодът:**
- Показва пълна информация за DataFrame
- Брой редове и колони
- Име, тип данни и брой non-null values за всяка колона
- Използвана памет

**Защо го правим:**
- Проверка за липсващи стойности (Non-Null Count)
- Проверка на типовете данни (int64, float64, object)
- Разбиране на структурата преди обработка

In [ ]:
# Задача 1.7: Покажете типовете данни на всяка колона
_____

## Част 2: Exploratory Data Analysis (10 минути)

In [ ]:
# Задача 2.1: Проверете за липсващи стойности (брой по колони)
print("Липсващи стойности:")
_____

**Какво прави кодът:**
- Създава DataFrame от True/False (True където има NaN)
- Брои True стойностите за всяка колона

**Защо го правим:**
- Липсващите стойности са ОГРОМЕН проблем в ML
- Трябва да ги открием преди обучение
- Има различни стратегии за справяне с тях

**Важна теория:**
- NaN = Not a Number (липсваща стойност)
- Причини: грешки при събиране, непопълнени полета, загубени данни
- Решения: изтриване, попълване със средна стойност, попълване с мода, прогнозиране

In [ ]:
# Задача 2.2: Покажете статистическо резюме на числовите колони
_____

**Важна теория - Статистиките:**
- **count**: Брой валидни стойности
- **mean**: Средна стойност (сума / брой)
- **std**: Стандартно отклонение (колко се разпръскват данните)
- **min/max**: Минимална/максимална стойност
- **25%/50%/75%**: Квартили (percentiles) - 50% е медианата

**Важни наблюдения:**
- price: огромен диапазон (1750000 до 13300000) - ще трябва scaling
- area: също голям диапазон (1650 до 16200)
- bedrooms: 1 до 6
- bathrooms: 1 до 4
- stories: 1 до 4
- parking: 0 до 3


In [ ]:
# Задача 2.3: Проверете уникалните стойности на колоните по категория
categorical_columns = ['mainroad', 'guestroom', 'basement', 'hotwaterheating',
                       'airconditioning', 'prefarea', 'furnishingstatus']

for col in categorical_columns:
    print(f"{col}: {_____}")

**Какво прави кодът:**
- Минава през всички колони
- Връща масив с уникалните стойности
- Печата ги за всяка колона

**Защо го правим:**
- Категориалните променливи трябва да се конвертират за ML
- Трябва да знаем какви стойности има всяка
- Помага да решим как да ги encode-нем

**Важни наблюдения:**
- **Binary колони** (2 стойности): mainroad, guestroom, basement, hotwaterheating, airconditioning, prefarea
  - Всички имат 'yes'/'no'
  - Може да се конвертират като 0/1 (Label Encoding)
- **Multi-class колона** (3+ стойности): furnishingstatus
  - Има 'furnished', 'semi-furnished', 'unfurnished'
  - Трябва One-Hot Encoding

In [ ]:
# Задача 2.3.5: Създайте нова колона с цени в милиони (за по-лесна визуализация)
df['price_millions'] = df['_____'] / _____
print("\nПървите 5 реда с цени в милиони:")
df[['price', 'price_millions']]._____()

**Защо го правим:**
- ВАЖНО за визуализациите!
- Оригиналните цени са 1,750,000 - 13,300,000 (1.75 - 13.3 милиона)
- Без тази трансформация histogram показва странни оси като "0.2", "1.2" (което всъщност означава 0.2×10⁷)
- След трансформация: осите показват "2", "4", "6" (милиони) - много по-интуитивно!

**Важна концепция - Data Transformation за Visualization:**
- Данните могат да се трансформират за по-добра визуализация БЕЗ да се променя смисълът
- Scaling за visualization ≠ Feature Scaling за ML
- Това е чисто за човешка четимост

In [ ]:
# Задача 2.4: Визуализирайте разпределението на цените (histogram с 50 bins)
_____(figsize=(10, 6))
_____(df['price_millions'], bins=_____, edgecolor='black', color='skyblue')
_____('Price (millions)')  # xlabel
_____('Frequency')  # ylabel
_____('Distribution of House Prices')  # title
_____(axis='y', alpha=0.3) #grid
_____()

In [ ]:
# Задача 2.5: Създайте correlation matrix за числовите колони
numeric_cols = ['price', 'area', 'bedrooms', 'bathrooms', 'stories', 'parking']
correlation_matrix = _____

# Визуализирайте с heatmap
_____(figsize=(10, 8))
_____(correlation_matrix, annot=_____, cmap='coolwarm', center=0)
_____('Correlation Matrix')
_____()

**Защо го правим:**
- Histogram показва разпределението на една променлива
- Виждаме дали данните са нормално разпределени
- Откриваме outliers и anomalies
- Разбираме концентрацията на стойностите

**Разлика - price vs price_millions:**
- **БЕЗ трансформация:** X-ос показва "0.2", "1.2", "2.2" което е ОБЪРКВАЩО
  - Matplotlib използва научна нотация и форматира като 0.2×10⁷ = 2,000,000
  - Учениците НЕ разбират какво означават тези числа!
- **С price_millions:** X-ос показва "2", "4", "6", "8", "10", "12" (милиони)
  - ИНТУИТИВНО и лесно за четене
  - Веднага се вижда че повечето къщи са между 2-8 милиона

**Важна теория - Histogram:**
- X-ос: стойности на променливата (price в милиони)
- Y-ос: честота (брой къщи в този ценови диапазон)
- Bins: интервали, в които групираме данните
- Повече bins = по-детайлна информация, но по-шумна

**Наблюдения:**
- Повечето къщи са в диапазона 3-7 милиона
- Малко скъпи къщи (10+ милиона) - дясна опашка
- Right-skewed distribution (положителна асиметрия)
- Няколко пика - може да има подгрупи (райони, типове къщи)

In [ ]:
# Задача 2.6: Направете boxplot за цена спрямо airconditioning
_____(figsize=(8, 6))
df._____(column='price_millions', by='airconditioning')
plt.ylabel('Price (millions)')
_____('Цена спрямо наличие на климатик')
plt.suptitle('')
_____()

In [ ]:
# Задача 2.7: Направете scatter plot на area vs price
_____(figsize=(10, 6))
_____(df['area'], df['price_millions'], alpha=0.5, color='coral')
_____('Area (sq ft)')
_____('Price (millions)')
_____('Area vs Price')
___.grid(alpha=0.3)
_____()

## Част 3: Обработка на категориални променливи (15 минути)

In [ ]:
# Задача 3.1: Създайте копие на DataFrame
df_processed = _____

In [ ]:
# Задача 3.2: Покажете първите 5 реда на binary колоните ПРЕДИ конвертиране
binary_columns = ['mainroad', 'guestroom', 'basement', 'hotwaterheating',
                  'airconditioning', 'prefarea']

print("Преди конвертиране:")
_____

In [ ]:
# Задача 3.3: Конвертирайте binary колони от yes/no в 1/0
for col in binary_columns:
    df_processed[col] = df_processed[col]._____(_____)  # map with dictionary

**Какво прави кодът:**
- Loop през всички binary колони
- `.map()` замества стойности по речник
- 'yes' става 1, 'no' става 0

**Защо го правим:**
- ML алгоритми работят САМО с числа
- Binary категории могат да се кодират като 0/1
- По-ефективно от One-Hot Encoding за binary данни

In [ ]:
# Задача 3.4: Покажете първите 5 реда СЛЕД конвертиране
print("След конвертиране:")
_____

**Какво прави кодът:**
- Показва същите колони СЛЕД transformation

**Защо го правим:**
- Проверка че mapping е успешен
- Визуална проверка преди/след
- Търсим грешки или missing values


In [ ]:
# Задача 3.5: Покажете разпределението на furnishingstatus ПРЕДИ One-Hot Encoding
print("Преди One-Hot Encoding:")
_____

**Защо го правим:**
- Проверка преди transformation
- Виждаме дали има imbalance в категориите
- Разбираме структурата на данните

**Важна теория:**
- furnishingstatus има 3 категории (multi-class)
- НЕ МОЖЕМ да използваме Label Encoding (furnished=0, semi=1, unfurnished=2)
- Защо не: Създава изкуствена подредба (unfurnished > semi > furnished)
- ML модел ще мисли че има ordinality (подредба)

In [ ]:
# Задача 3.6: Приложете One-Hot Encoding на furnishingstatus колоната
# Използвайте drop_first=True за да избегнете dummy variable trap
df_processed = _____

**Важна теория - One-Hot Encoding:**
- Превръща една multi-class колона в множество binary колони
- Всяка категория става отделна колона с 0/1
- Пример:
  - Преди: furnishingstatus = 'furnished'
  - След: furnishingstatus_semi-furnished = 0, furnishingstatus_unfurnished = 0
  - (furnished се разбира от факта че и двете са 0)

**Защо drop_first=True:**
- Dummy Variable Trap!
- Ако имаме 3 категории, ни трябват САМО 2 колони
- Третата категория се разбира от останалите
- Пример: ако semi=0 и unfurnished=0, то е furnished
- Без drop_first: създаваме perfect multicollinearity (проблем за ML)

**Какво създава:**
- furnishingstatus колоната се премахва
- Създават се нови колони:
  - furnishingstatus_semi-furnished (0 или 1)
  - furnishingstatus_unfurnished (0 або 1)
- furnished се представя като (0, 0)

In [ ]:
# Задача 3.7: Покажете имената на всички колони СЛЕД One-Hot Encoding
print("След One-Hot Encoding:")
_____

**Защо го правим:**
- Проверка че One-Hot Encoding е успешен
- Виждаме новите имена на колоните
- Преброяваме общия брой колони

In [ ]:
# Задача 3.8: Проверете новия размер на данните
print(f"Нов размер: {_____}")

**Какво прави кодът:**
- Показва новия размер на DataFrame

**Защо го правим:**
- Проверка че броят колони се е променил правилно
- Броят редове трябва да е същият (545)
- Броят колони е нараснал с 1 (13 -> 14)

## Част 4: Feature Scaling

In [ ]:
# Задача 4.1: Импортирайте StandardScaler
from sklearn._____ import _____

**Защо го правим:**
- StandardScaler прави стандартизация (standardization)
- Превръща всички features в една и съща скала

**Важна теория - Feature Scaling:**
- ML алгоритми "виждат" големите числа като по-важни
- Пример: area (1000-16000) vs bedrooms (1-6)
- Без scaling: area ще доминира, bedrooms ще се игнорира
- Scaling прави всички features "равнопоставени"

**Съвети за преподаване:**
- Обяснете защо scaling е критичен
- Споменете че не всички алгоритми го изискват (tree-based не се нуждаят)
- Но за linear models, neural networks, SVM е ЗАДЪЛЖИТЕЛНО

In [ ]:
# Задача 4.2: Създайте копие на данните за scaling
df_scaled = _____

**Защо го правим:**
- Можем да сравним преди/след scaling
- Добра практика за experimentation

In [ ]:
# Задача 4.3: Изберете колони за scaling (всички освен 'price')
features_to_scale = df_scaled._____(_____, axis=_____).columns

**Защо НЕ scale-ваме price:**
- price е нашата target променлива (y)
- Scale-ваме само features (X)
- Target може да остане в оригинална скала (по-лесно интерпретиране)

In [ ]:
# Задача 4.4: Създайте StandardScaler обект
scaler = _____

**Важна теория - StandardScaler:**
- Прави Z-score standardization
- Формула: z = (x - mean) / std
- Резултат: mean = 0, std = 1
- Всяка feature получава същото разпределение

**Алтернативи:**
- MinMaxScaler: скалира в диапазон [0, 1]
- RobustScaler: устойчив на outliers
- StandardScaler е най-често използван

In [ ]:
# Задача 4.5: Приложете fit_transform на избраните features
df_scaled[features_to_scale] = scaler._____(df_scaled[features_to_scale])

**Какво прави кодът:**
- ДВЕ неща:
  1. **fit()**: Изчислява mean и std от данните
  2. **transform()**: Прилага scaling с тези параметри
- Резултатът се записва обратно в df_scaled

**Важна теория - fit vs transform:**
- **fit()**: "Учи" от данните (изчислява mean, std)
- **transform()**: Прилага научените параметри
- **fit_transform()**: fit + transform заедно (за train data)

**КРИТИЧНО ВАЖНО за train/test:**
- Train data: fit_transform() (учи И прилага)
- Test data: само transform() (прилага СЪЩИТЕ параметри от train)
- НИКОГА не правим fit на test data! (data leakage)

In [ ]:
# Задача 4.6: Сравнете 3 реда ПРЕДИ и СЛЕД scaling
print("Преди scaling:")
print(df_processed[['area', 'bedrooms', 'bathrooms']]._____(3))

print("\nСлед scaling:")
print(df_scaled[['area', 'bedrooms', 'bathrooms']]._____(3))

**Защо го правим:**
- Визуална демонстрация на ефекта от scaling
- Проверка че scaling е приложен правилно

**Какво да забележим:**
- Преди: area е ~7000, bedrooms е ~3
- След: всички около 0, в диапазон [-3, 3]
- Негативни стойности са нормални (по-малко от средното)
- Положителни стойности = по-големи от средното


## Част 5: Train/Test Split

In [ ]:
# Задача 5.1: Импортирайте train_test_split
from sklearn._____ import _____

**Защо го правим:**
- Train/Test split е ФУНДАМЕНТАЛНА концепция в ML
- Разделяме данните за обучение и тестване

**Важна теория - Train/Test Split:**
- **Проблем**: Ако тестваме на същите данни, с които обучаваме - модела запомня (overfitting)
- **Решение**: Разделяме данните на две части
  - **Train set**: За обучение на модела (обикновено 70-80%)
  - **Test set**: За оценка на модела (обикновено 20-30%)
- Модела НИКОГА не вижда test set по време на обучение


In [ ]:
# Задача 5.2: Разделете данните на X (features) и y (target)
X = df_scaled._____(_____, axis=_____)
y = df_scaled[_____]

**Какво прави кодът:**
- **X**: Всички колони ОСВЕН price (features, independent variables)
- **y**: Само price колоната (target, dependent variable)

**Защо го правим:**
- ML модели имат формата: y = f(X)
- X = входните данни (characteristics на къщата)
- y = изходът който искаме да предскажем (цената)

**Терминология:**
- **X**: Features, predictors, independent variables, input
- **y**: Target, label, dependent variable, output

In [ ]:
# Задача 5.3: Проверете размерите на X и y
print(f"X size: {_____}")
print(f"y size: {_____}")

In [ ]:
# Задача 5.4: Разделете на train и test sets (80/20 split, random_state=42)
X_train, X_test, y_train, y_test = _____(X, y, test_size=_____, random_state=_____)

**Параметри:**
- **test_size=0.2**: 20% за test, 80% за train
  - Може да е float (0.2 = 20%) или int (100 = 100 samples)
  - Стандарт: 0.2 или 0.25
- **random_state=42**: Seed за random number generator
  - Гарантира reproducibility (същото разделяне всеки път)
  - 42 е популярен (от "The Hitchhiker's Guide to the Galaxy")

**Как работи:**
1. Разбърква данните (shuffle)
2. Взима първите 80% за train
3. Взима последните 20% за test
4. Прави същото за X и y (запазва съответствието)

**Важна теория:**
- Train и test са НАПЪЛНО РАЗДЕЛЕНИ
- Няма припокриване между тях
- Test set е "frozen" - не го пипаме до финалната оценка

In [ ]:
# Задача 5.5: Проверете размерите на train и test sets
print(f"Train set - X: {_____}, y: {_____}")
print(f"Test set - X: {_____}, y: {_____}")

**Изчисления:**
- Total: 545 къщи
- Train: 80% от 545 = 436 къщи
- Test: 20% от 545 = 109 къщи
- X има 13 features и в двата случая

In [ ]:
# Задача 5.6: Покажете статистика на y_train
print("Train set price statistics:")
_____

In [ ]:
# Задача 5.7: Покажете статистика на y_test
print("Test set price statistics:")
_____

## Част 6: Изграждане на Simple Linear Regression модел

In [ ]:
# Задача 6.1: Импортирайте LinearRegression и метриките
from sklearn.linear_model import _____
from sklearn.metrics import _____, _____, _____

In [ ]:
# Задача 6.2: Създайте нови X_train и X_test само с 'area' колоната
# Забележка: Използваме scaled данните, но само 'area' feature
X_train_simple = X_train[['_____']]
X_test_simple = X_test[['_____']]

print(f"X_train_simple shape: {X_train_simple.shape}")
print(f"X_test_simple shape: {X_test_simple.shape}")

In [ ]:
# Задача 6.3: Създайте Linear Regression модел
model = _____()

In [ ]:
# Задача 6.4: Обучете модела с тренировъчните данни
model._____(X_train_simple, y_train)

print("✓ Моделът е обучен успешно!")

## Част 7: Анализ на коефициентите

In [ ]:
# Задача 7.1: Извлечете slope (коефициент) и intercept
slope = model._____[0]
intercept = model._____

print(f"Slope (коефициент): {slope:,.2f}")
print(f"Intercept: {intercept:,.2f}")
print(f"\nУравнение: price = {slope:,.2f} × area + {intercept:,.2f}")

**Важна теория - Интерпретация:**
- **Slope (m)**: Колко се променя y при промяна на x с 1 unit
  - Ако slope = 1,500,000: При +1 unit area → +1,500,000 price
  - ВАЖНО: В нашия случай area е SCALED (z-score)
  - Така че slope показва: +1 std на area → +slope change в price
- **Intercept (b)**: Прогнозираната цена когато area = 0
  - В scaled данни: когато area = средната стойност
  - Intercept ≈ средната цена на train set

**КРИТИЧНО - Scaled Features:**
- Area е standardized (mean=0, std=1)
- Slope НЕ е "цена на sq ft" в оригинални единици!
- За интерпретация в оригинални единици:
  - Трябва да de-scale slope: slope_original = slope / scaler.scale_[area_index]
  - Или да обучим модел на unscaled данни (но scaling е добра практика)

**Математически детайли:**
- `coef_` е numpy array (може да има много slopes за много features)
- За Simple Linear Regression: len(coef_) = 1
- За Multiple Linear Regression: len(coef_) = брой features

**Очаквани стойности:**
- Slope: Положителен (area ↑ → price ↑)
- Slope величина: Зависи от scaling, обикновено в хиляди
- Intercept: Около средната цена (~4-5 милиона)

In [ ]:
# Задача 7.2: Интерпретирайте коефициента
print(f"Интерпретация:")
print(f"За всеки допълнителен стандартизиран unit от площ,")
print(f"цената на жилището се увеличава средно с {slope:,.2f} единици.")

## Част 8: Прогнозиране

In [ ]:
# Задача 8.1: Направете прогнози на test set
y_pred = model._____(X_test_simple)

print("Първите 5 прогнози:")
print(y_pred[:5])

**Как работи:**
- За всеки ред в X_test_simple:
  - y_pred = slope * area + intercept
- Това е просто прилагане на формулата

**ВАЖНО:**
- НЕ ПРОМЕНЯ модела
- Използва СЪЩИТЕ slope/intercept от fit()
- Можем да извикаме predict() много пъти

In [ ]:
# Задача 8.2: Сравнете първите 5 прогнози с действителните стойности
comparison = pd.DataFrame({
    'Actual': y_test[:5].values,
    'Predicted': y_pred[:5],
    'Difference': y_test[:5].values - y_pred[:5]
})
print(comparison)

## Част 9: Евалуация на модела

In [ ]:
# Задача 9.1: Изчислете Mean Squared Error (MSE)
mse = _____(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse:,.2f}")

**Какво прави кодът:**
- Изчислява MSE (Mean Squared Error)
- Формула: MSE = (1/n) * Σ(actual - predicted)²

**Защо MSE:**
- Наказва големите грешки повече (заради квадрата)
- Винаги положителна стойност
- По-малка MSE = по-добър модел

**Недостатък на MSE:**
- Единиците са квадратни (price² !?)
- Трудно за интерпретация
- Затова използваме RMSE (квадратен корен от MSE)

In [ ]:
# Задача 9.2: Изчислете Root Mean Squared Error (RMSE)
rmse = np._____(mse)
print(f"Root Mean Squared Error (RMSE): {rmse:,.2f}")
print(f"Интерпретация: Средната грешка в прогнозата е ±{rmse:,.0f} единици")

**Какво прави кодът:**
- RMSE = √MSE
- Връща грешката в оригиналните единици

**Защо RMSE:**
- ИНТЕРПРЕТИРУЕМ! Същите единици като target
- "В среден случай, грешим с ±RMSE"
- По-малка RMSE = по-добър модел

**Интерпретация:**
- RMSE = 1,200,000 означава: "Средно грешим с ±1.2 милиона"
- Сравнете с диапазона на price (1.75M - 13.3M)
- Ако RMSE е малка спрямо диапазона → добър модел


In [ ]:
# Задача 9.3: Изчислете Mean Absolute Error (MAE)
mae = _____(y_test, y_pred)
print(f"Mean Absolute Error (MAE): {mae:,.2f}")

**Какво прави кодът:**
- MAE = (1/n) * Σ|actual - predicted|
- Средната абсолютна грешка

**Защо MAE:**
- По-устойчив към outliers от RMSE
- НЕ наказва големи грешки толкова силно
- Интуитивна интерпретация

**MAE vs RMSE:**
- MAE винаги ≤ RMSE
- Ако RMSE >> MAE → има outliers с големи грешки
- Ако RMSE ≈ MAE → грешките са равномерни


In [ ]:
# Задача 9.4: Изчислете R² Score
r2 = _____(y_test, y_pred)
print(f"R² Score: {r2:.4f}")
print(f"Интерпретация: Моделът обяснява {r2*100:.2f}% от вариацията в цените")

**Какво прави кодът:**
- R² (R-squared, coefficient of determination)
- Най-важната метрика за regression!

**Формула и интерпретация:**
- R² = 1 - (SS_res / SS_tot)
- SS_res = Σ(actual - predicted)² (residual sum of squares)
- SS_tot = Σ(actual - mean)² (total sum of squares)
- **R² показва: Каква част от вариацията е обяснена от модела**

**Диапазон на стойности:**
- **R² = 1.0**: Перфектен модел (100% обяснена вариация)
- **R² = 0.8**: Много добър модел (80% обяснена вариация)
- **R² = 0.5**: Среден модел (50% обяснена вариация)
- **R² = 0.0**: Моделът не е по-добър от средната стойност
- **R² < 0**: Моделът е по-лош от средната стойност (много лошо!)

**За Simple Linear Regression с area:**
- Защо не по-високо: Само 1 feature (area)
- Price зависи и от: location, bedrooms, condition, etc.
- С всички features → R² може да е 0.7-0.8+

**КРИТИЧНО ВАЖНО за преподаване:**
- R² НЕ Е "процент на точност"!
- R² е "процент на обяснена вариация"
- R² = 0.5 НЕ означава "50% от прогнозите са точни"
- R² = 0.5 означава "модела обяснява 50% от разликите в цените"

**Пример:**
- Защо една къща струва 8M, а друга 4M?
- R² = 0.5 означава: "50% от тази разлика се обяснява с area"
- Останалите 50% се обясняват с други фактори (не в модела)

In [ ]:
# Задача 9.5: Обобщете всички метрики
print("\n" + "="*50)
print("ОБОБЩЕНИЕ НА МЕТРИКИТЕ")
print("="*50)
print(f"MSE:  {mse:,.2f}")
print(f"RMSE: {rmse:,.2f}")
print(f"MAE:  {mae:,.2f}")
print(f"R²:   {r2:.4f} ({r2*100:.2f}%)")
print("="*50)

## Част 10: Визуализация

In [ ]:
# Задача 10.1: Визуализирайте regression line върху test данните
plt.figure(figsize=(12, 5))

# Subplot 1: Scatter plot с regression line
plt.subplot(1, 2, 1)
plt.scatter(X_test_simple, y_test, alpha=0.5, color='blue', label='Действителни цени')
plt.plot(X_test_simple, y_pred, color='red', linewidth=2, label='Regression line')
plt.xlabel('Area (scaled)')
plt.ylabel('Price')
plt.title('Linear Regression: Area vs Price')
plt.legend()
plt.grid(alpha=0.3)

# Subplot 2: Predicted vs Actual
plt.subplot(1, 2, 2)
plt.scatter(y_test, y_pred, alpha=0.5, color='green')
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()],
         'r--', linewidth=2, label='Perfect prediction')
plt.xlabel('Действителни цени')
plt.ylabel('Прогнозирани цени')
plt.title('Predicted vs Actual Prices')
plt.legend()
plt.grid(alpha=0.3)

plt.tight_layout()
plt.show()

## Част 11: Анализ на остатъците - Residual Analysis



In [ ]:
# Задача 11.1: Изчислете остатъците (residuals = actual - predicted)
residuals = y_test - _____

print(f"Първите 10 остатъка:")
print(residuals[:10].values)

In [ ]:
# Задача 11.2: Визуализирайте остатъците
plt.figure(figsize=(14, 5))

# Subplot 1: Residual plot
plt.subplot(1, 3, 1)
plt.scatter(y_pred, residuals, alpha=0.5, color='purple')
plt.axhline(y=0, color='red', linestyle='--', linewidth=2)
_____('Прогнозирани цени')   #xlabel
_____('Остатъци (Residuals)') #ylabel
_____('Residual Plot') #title
_____(alpha=0.3) #grid

# Subplot 2: Histogram на остатъците (bins=30, цвят по избор)
plt.subplot(1, 3, 2)
_____(residuals, bins=____, edgecolor='black', color=____, alpha=0.7) #hist
_____('Остатъци') #xlabel
_____('Честота')  #ylabel
_____('Разпределение на остатъците') #title
_____(axis='y', alpha=0.3) #grid

# Subplot 3: Q-Q plot (за проверка на нормалност)
plt.subplot(1, 3, 3)
from scipy import stats
stats.probplot(residuals, dist="norm", plot=plt)
_____('Q-Q Plot') #title
_____(alpha=0.3) #grid

plt.tight_layout()
_____() #show

**Графика 1 - Regression Line:**
- Показва оригиналните данни (сини точки)
- Червена линия = модела (y = mx + b)
- Разстоянието точка-линия = грешка (residual)
- Добър модел: точките близо до линията

**Графика 2 - Predicted vs Actual:**
- X-ос: Истинските цени
- Y-ос: Прогнозите
- Червена пунктирана: перфектна прогноза
- Точки на линията = перфектна прогноза
- Точки под линията = подценени (модела прогнозира по-ниско)
- Точки над линията = надценени (модела прогнозира по-високо)

**Какво да търсим:**
- Разсейване около линията = грешки
- Patterns в разсейването = проблем (трябва random)
- Outliers = точки далеч от линията
- И двете графики показват СЪЩОТО, но по различен начин
- Графика 1: Как модела fit-ва данните
- Графика 2: Колко добри са прогнозите

**Очакван резултат:**
- График 1: Възходяща линия, точки около нея
- График 2: Точки около диагонала, някои outliers

In [ ]:
# Задача 11.3: Анализирайте остатъците
print("Статистика на остатъците:")
print(f"Средна стойност: {residuals.mean():,.2f} (трябва да е близо до 0)")
print(f"Стандартно отклонение: {residuals.std():,.2f}")
print(f"Min: {residuals.min():,.2f}")
print(f"Max: {residuals.max():,.2f}")

## ✅ Module 1A Завършен!

### Научихте:
- ✓ Зареждане и инспектиране на данни
- ✓ Exploratory Data Analysis (EDA)
- ✓ Визуализации (histogram, boxplot, scatter, heatmap)
- ✓ Обработка на binary categorical variables (Label Encoding)
- ✓ One-Hot Encoding за multi-class категории
- ✓ Feature Scaling със StandardScaler
- ✓ Train/Test Split

### Следващ модул:
**Module 1B:** Titanic Dataset - Missing Values & Feature Engineering